pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,459 @@
+	"github.com/packethost/packngo"
+
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/cloud"
+	cloudprovidererrors "github.com/kubermatic/machine-controller/pkg/cloudprovider/errors"
+	"github.com/kubermatic/machine-controller/pkg/cloudprovider/instance"
+	"github.com/kubermatic/machine-controller/pkg/providerconfig"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
+	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+const (
+	machineUIDTag       = "kubermatic-machine-controller:machine-uid"
+	defaultBillingCycle = "hourly"
+)
+
+// New returns a Packet provider
+func New(configVarResolver *providerconfig.ConfigVarResolver) cloud.Provider {
+	return &provider{configVarResolver: configVarResolver}
+}
+
+type RawConfig struct {
+	APIKey       providerconfig.ConfigVarString   `json:"apiKey"`
+	ProjectID    providerconfig.ConfigVarString   `json:"projectID"`
+	BillingCycle providerconfig.ConfigVarString   `json:"billingCycle"`
+	InstanceType providerconfig.ConfigVarString   `json:"instanceType"`
+	Facilities   []providerconfig.ConfigVarString `json:"facilities"`
+	OS           providerconfig.ConfigVarString   `json:"os"`
+	Tags         []providerconfig.ConfigVarString `json:"tags"`
+}
+
+type Config struct {
+	APIKey       string
+	ProjectID    string
+	BillingCycle string
+	InstanceType string
+	Facilities   []string
+	OS           string
+	Tags         []string
+}
+
+type provider struct {
+	configVarResolver *providerconfig.ConfigVarResolver
+}
+
+func (p *provider) getConfig(s v1alpha1.ProviderSpec) (*Config, *RawConfig, *providerconfig.Config, error) {
+	if s.Value == nil {
+		return nil, nil, nil, fmt.Errorf("machine.spec.providerconfig.value is nil")
+	}
+	pconfig := providerconfig.Config{}
+	err := json.Unmarshal(s.Value.Raw, &pconfig)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	rawConfig := RawConfig{}
+	if err = json.Unmarshal(pconfig.CloudProviderSpec.Raw, &rawConfig); err != nil {
+		return nil, nil, nil, err
+	}
+
+	c := Config{}
+	c.APIKey, err = p.configVarResolver.GetConfigVarStringValueOrEnv(rawConfig.APIKey, "PACKET_API_KEY")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"apiKey\" field, error = %v", err)
+	}
+	c.ProjectID, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.ProjectID)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"projectID\" field, error = %v", err)
+	}
+	c.InstanceType, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.InstanceType)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"instanceType\" field, error = %v", err)
+	}
+	c.BillingCycle, err = p.configVarResolver.GetConfigVarStringValue(rawConfig.BillingCycle)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to get the value of \"billingCycle\" field, error = %v", err)
+	}
+	if c.BillingCycle == "" {
+		c.BillingCycle = defaultBillingCycle
```

Comment:
Defaulting should potentially be moved into `AddDefaults`.

Have a look at the AWS code for a reference: https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/aws/provider.go#L299

Reply:
I originally had it in `AddDefaults()`, but that doesn't get exercised in the same way. I am fine with moving it into `AddDefaults()` _also_, and then having a single internal private function that validates and adds defaults, and that one gets called _both_ by `AddDefaults()` _and_ by the private `getConfig()`, if you prefer that. But I definitely want it here as well.

Let me know.

Reply:
Having it both places sounds good :+1:

Reply:
Was a bit of a pain, but got it in. I don't love the duplication of code, but until this becomes easier, as you mentioned in the Slack, this will do.